File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.goodNodes = 0

        def dfs(root, minNumber):
            if root:
                if root.val >= minNumber:
                    self.goodNodes += 1
                dfs(root.left, max(minNumber, root.val))
                dfs(root.right, max(minNumber, root.val))
        dfs(root, float('-inf'))
        return self.goodNodes

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        } else {
            if (root.val == p.val || root.val == q.val) {
                return root;
            } else if (root.val > p.val && root.val > q.val) {
                TreeNode answer = lowestCommonAncestor(root.left, p, q);
                if (answer != null) {
                    return answer;
                } else {
                    return root;
                }
            } else if (root.val < p.val && root.val < q.val) {
                TreeNode answer = lowestCommonAncestor(root.right, p, q);
                if (answer != null) {
                    return answer;
                } else {
                    return root;
                }
            } else {
                return root;
            }
        }
    }
}
